nexus/db-queries/src/db/datastore/deployment.rs

```diff
@@ -2358,93 +2445,119 @@ mod tests {
-            }
-        });
+        let conn = datastore
+            .pool_connection_authorized(&opctx)
+            .await
+            .expect("failed to get connection");
 
-        // None of our spawned tasks should be able to make progress:
-        // `ensure_resources_task` is waiting for us to set
-        // `return_on_completion` to true, and the other two should be
-        // queued by Cockroach, because
-        // `blueprint_ensure_external_networking_resources` should have
-        // performed a `SELECT ... FOR UPDATE` on the current target, forcing
-        // the query that wants to change it to wait until the transaction
-        // completes.
+        // NOTE: Performing this "SELECT" is a necessary step for our test, even
+        // though we don't actually care about the result.
         //
-        // We'll somewhat haphazardly test this by trying to wait for any
-        // task to finish, and succeeding on a timeout of a few seconds. This
-        // could spuriously succeed if we're executing on a very overloaded
-        // system where we hit the timeout even though one of the tasks is
-        // actually making progress, but hopefully will fail often enough if
-        // we've gotten this wrong.
-        tokio::select! {
-            result = &mut ensure_resources_task => {
-                panic!(
-                    "unexpected completion of \
-                     `blueprint_ensure_external_networking_resources`: \
-                     {result:?}",
-                );
-            }
-            result = &mut update_target_task => {
-                panic!(
-                    "unexpected completion of \
-                     `blueprint_target_set_current`: {result:?}",
-                );
-            }
-            result = &mut current_target_task => {
-                panic!(
-                    "unexpected completion of \
-                     `blueprint_target_get_current`: {result:?}",
-                );
-            }
-            _ = tokio::time::sleep(Duration::from_secs(5)) => (),
-        }
+        // If we don't perform this read, it is possible for CockroachDB to
+        // logically order the "Ensure Resources" task before the blueprint
+        // target changes.
+        //
+        // More on this in the block comment below.
+        let _external_ips = QueryBuilder::new()
+            .sql("SELECT id FROM omicron.public.external_ip WHERE time_deleted IS NULL")
+            .query::<diesel::sql_types::Uuid>()
+            .load_async::<uuid::Uuid>(&*conn)
+            .await
+            .expect("SELECT external IPs");
 
-        // Release `ensure_resources_task` to finish.
-        return_on_completion.store(true, Ordering::SeqCst);
+        // == WHAT ORDERING DO WE EXPECT?
+        //
+        // We expect to have the following reads/writes:
+        //
+        // | Ensure Resources |  Set current target  |
```

Comment:
Maybe 
```suggestion
        // | Ensure Resources |  Unit test  |
```

since the `R(data)` isn't really part of setting the target?

Reply:
Sounds good. This was the old "task name" from when we had three tasks here, but I agree it's clearer to just state "Unit test".

(Also, hopefully the ordering is more clear / controlled by calling this inline, rather than punting all work to multiple `tokio::task`s)

Reply:
Yeah, it definitely is :+1: